---
title: "Adamson smallRNA - pegRNA"
output:
  html_notebook:
    toc: true
    code_folding: hide
---

## Load libraries

This project uses [`renv`](https://rstudio.github.io/renv/articles/renv.html)
to keep track of installed packages. Install `renv` if not installed and load
dependencies with `renv::restore()`.

```r
install.packages("renv")
renv::restore()
```

```{r message=FALSE}
library(readr)
library(dplyr)
library(tidyr)
library(reshape2)
library(GenomicRanges)
library(pheatmap)
library(tibble)
library(ggplot2)
library(stringr)
library(cowplot)
library(markdown)
library(RColorBrewer)
library(GenomicAlignments)
```

## Read data

1. Get list of samples

```{r}
samples <- read_tsv("config/samples.tsv", show_col_types = FALSE)
units <- read_tsv("config/units.tsv", show_col_types = FALSE)
sample_units <- dplyr::left_join(samples, units, by = "sample_name") %>%
  unite(sample_unit, sample_name, unit_name, remove = FALSE)
sample_units
```

2. Read Samtools `idxstats` to get human coverage for normalization

Notes:
* The counts include the total number of reads aligned, they 
  are not limited to uniquely aligned reads.
* The counts are reads, not pairs or fragments

```{r}
idxstats_exogenousrna_dir <-
  "results/samtools_idxstats/exogenous_rna/"

idxstats_human_dir <-
  "results/samtools_idxstats/Homo_sapiens.GRCh38.dna.primary_assembly/"

bowtie2_human_logs <-
  "results/logs/bowtie2/Homo_sapiens.GRCh38.dna.primary_assembly/"

idxstats <- tibble(
  sample = sample_units$sample_unit,
  exogenous_rna_mapped_reads = NaN,
  grch38_mapped_reads = NaN,
  unmapped_reads = NaN
)

for (row in seq_len(nrow(sample_units))) {
  sample <- sample_units[row, ]$sample_unit

  # Read `idsxstats` for exogenous mapped reads
  exogenous_rna_stats <- read_tsv(
    file.path(idxstats_exogenousrna_dir, sprintf("%s.bam.idxstats", sample)),
    col_names = c(
      "sequence_name", "sequence_length",
      "mapped_reads", "unmapped_reads"
    ),
    col_types = "ciii"
  )
  exogenous_rna_mapped_reads <- exogenous_rna_stats %>%
    filter(!sequence_name %in% c("*")) %>%
    select(mapped_reads) %>%
    sum()

  # Read `idxstats` for human mapped reads
  human_stats <- read_tsv(
    file.path(idxstats_human_dir, sprintf("%s.bam.idxstats", sample)),
    col_names = c(
      "sequence_name", "sequence_length",
      "mapped_reads", "unmapped_reads"
    ),
    col_types = "ciii"
  )
  grch38_mapped_reads <- human_stats %>%
    filter(!sequence_name %in% c("*")) %>%
    select(mapped_reads) %>%
    sum()

  # Read bowtie2 logs for unmapped reads
  bowtie2_log <- readLines(
    file.path(bowtie2_human_logs, sprintf("%s.log", sample))
  )
  total_pairs <- strtoi(str_split(bowtie2_log[1], " ")[[1]][1])
  total_reads <- total_pairs * 2
  unmapped_reads <- total_reads - grch38_mapped_reads

  # Consolidate counts for rows
  idxstats <- rows_update(idxstats,
    tibble(
      sample = sample,
      exogenous_rna_mapped_reads = exogenous_rna_mapped_reads,
      grch38_mapped_reads = grch38_mapped_reads,
      unmapped_reads = unmapped_reads,
    ),
    by = "sample"
  )
}
idxstats
```

3. Read `bedpe` files to get exogenous rna coverage of paired reads

```{r}
bedpe_data <- tibble()
for (sample in sample_units$sample_unit) {
  data <-
    readr::read_tsv(
      sprintf(
        "results/alignments/exogenous_rna/bedpe/%s.bedpe", sample
      ),
      col_names = c(
        "chrom1", "chrom1Start", "chrom1End",
        "chrom2", "chrom2Start", "chrom2End",
        "name", "score", "strand1", "strand2"
      ),
      col_types = "ciiciicicc"
    )
  bedpe_data <- tibble(rbind(
    bedpe_data,
    cbind(
      filter(idxstats, sample == !!sample),
      data
    )
  ))
}
bedpe_data
```

## Exogenous RNA annotation data

The exogenous RNA references have the following structure

**PJY103**

| start | end | description |
|:--- |:--- |:--- |
| 1 | 330 | plasmid-1 |
| 331 | 351 | protospacer |
| 352 | 427 | gRNA scaffold |
| 428 | 453 | extension |
| 454 | 460 | terminator |
| 461 | 2306 | plasmid-2 |

Setup annotation for exogenous RNA regions

```{r}
annotation_pjy103 <- tibble(chrom2End = 1:2306, label = "plasmid")
annotation_pjy103[331:351, ]$label <- "protospacer"
annotation_pjy103[352:427, ]$label <- "gRNA scaffold"
annotation_pjy103[428:453, ]$label <- "extension"
annotation_pjy103[454:460, ]$label <- "terminator"
annotation_pjy103 <- annotation_pjy103 %>%
  mutate(region = paste0("PJY103_", chrom2End)) %>%
  select(-chrom2End) %>%
  column_to_rownames("region")
```

**PJY300**

| start | end | description |
|:--- |:--- |:--- |
| 1 | 330 | plasmid-1 |
| 331 | 351 | protospacer |
| 352 | 427 | gRNA scaffold |
| 428 | 453 | extension |
| 454 | 490 | riboswitch |
| 491 | 497 | terminator |
| 498 | 2343 | plasmid-2 |

```{r}
annotation_pjy300 <- tibble(chrom2End = 1:2306, label = "plasmid")
annotation_pjy300[331:351, ]$label <- "protospacer"
annotation_pjy300[352:427, ]$label <- "gRNA scaffold"
annotation_pjy300[428:453, ]$label <- "extension"
annotation_pjy300[454:490, ]$label <- "riboswitch"
annotation_pjy300[491:497, ]$label <- "terminator"
annotation_pjy300 <- annotation_pjy300 %>%
  mutate(region = paste0("PJY300_", chrom2End)) %>%
  select(-chrom2End) %>%
  column_to_rownames("region")
```

## Coverage

### Concordant vs Discordant paired reads

Concordant pairs are pairs of reads that:

* Align on the same pegRNA
* Align within 500 bp of each other
* Align in the expected forward-reverse orientation (`--> .. <--`)

Discordant reads aligned but whose mate:

* Did not align (on the pegRNA)
* Aligned more than 500 bp away
* Aligned in an unexpected orientation

```{r fig.width=10, fig.height=10}
## Config and function definition

bam_dir <- "results/alignments/exogenous_rna/sorted"

param_discordant <- ScanBamParam(flag = scanBamFlag(isProperPair = FALSE))
param_concordant <- ScanBamParam(flag = scanBamFlag(isProperPair = TRUE))
last_day <- 0

concordant_cell_line_colors <- list(
  "Parental" = "#de425b",
  "P1E10" = "#488f31"
)

discordant_cell_line_colors <- list(
  "Parental" = "#e98e6a",
  "P1E10" = "#93b168"
)

pegrna_plots <- function(rna_species, normalization_factor, ylim, ylab) {
  par(mfrow = c(2, 1))

  for (s in (sample_units %>%
    arrange(exogenous_rna, day, cell_line) %>%
    filter(exogenous_rna == rna_species))$sample_unit) {
    exogenous_mapped_reads <- bedpe_data %>%
      filter(sample == s) %>%
      select(.data[[normalization_factor]]) %>%
      unique() %>%
      pull(1)

    cell_line <- sample_units %>%
      filter(sample_unit == s) %>%
      pull(cell_line)
    day <- sample_units %>%
      filter(sample_unit == s) %>%
      pull(day)

    discordant <- readGAlignments(sprintf("%s/%s.bam", bam_dir, s),
      param = param_discordant
    )
    cov_discordant <- coverage(discordant)
    cov_discordant_norm_pegrna <- cov_discordant / exogenous_mapped_reads

    df_complete <- bedpe_data %>%
      filter(sample == s)
    concordant <- GRanges(
      ranges = IRanges(
        start = df_complete$chrom1Start,
        end = df_complete$chrom2End
      ),
      seqnames = df_complete$chrom1
    )
    cov_concordant <- coverage(concordant)
    cov_concordant_norm_pegrna <- cov_concordant / exogenous_mapped_reads

    if (day != last_day) {
      plot(cov_concordant_norm_pegrna[[rna_species]],
        type = "l",
        col = concordant_cell_line_colors[[cell_line]],
        ylim = ylim,
        xlim = c(0, 600),
        main = sprintf("Day %s", day),
        xlab = sprintf("%s position", rna_species),
        ylab = ylab
      )
      lines(cov_discordant_norm_pegrna[[rna_species]],
        type = "l",
        col = discordant_cell_line_colors[[cell_line]]
      )
    } else {
      lines(cov_concordant_norm_pegrna[[rna_species]],
        type = "l",
        col = concordant_cell_line_colors[[cell_line]]
      )
      lines(cov_discordant_norm_pegrna[[rna_species]],
        type = "l",
        col = discordant_cell_line_colors[[cell_line]]
      )
    }
    legend("topleft",
      legend = c(
        "Parental:Concordant",
        "P1E10:Concordant",
        "Parental:Discordant",
        "P1E10:Discordant"
      ),
      col = unlist(c(
        concordant_cell_line_colors,
        discordant_cell_line_colors
      )),
      lty = 1,
      cex = 1.0
    )
    last_day <- day
  }
}
```

### Normalized to pegRNA reads

#### PJY103 (pegRNA)

```{r fig.width=10, fig.height=10}
pegrna_plots(
  rna_species = "PJY103",
  normalization_factor = "exogenous_rna_mapped_reads",
  ylim = c(0, 0.5),
  ylab = "Coverage (normalized to pegRNA reads)"
)
```
#### PJY300 (epegRNA)

```{r fig.width=10, fig.height=10}
pegrna_plots(
  rna_species = "PJY300",
  normalization_factor = "exogenous_rna_mapped_reads",
  ylim = c(0, 0.5),
  ylab = "Coverage (normalized to pegRNA reads)"
)
```


### Normalized to Human mapped reads

#### PJY103 (pegRNA)

```{r fig.width=10, fig.height=10}
pegrna_plots(
  rna_species = "PJY103",
  normalization_factor = "grch38_mapped_reads",
  ylim = c(0, 0.0012),
  ylab = "Coverage (normalized to Human reads)"
)
```
#### PJY300 (epegRNA)

```{r fig.width=10, fig.height=10}
pegrna_plots(
  rna_species = "PJY300",
  normalization_factor = "grch38_mapped_reads",
  ylim = c(0, 0.004),
  ylab = "Coverage (normalized to Human reads)"
)
```
### Normalize to reads mapped to 5.8S rRNA

[5.8S rRNA in Ensembl assembly](http://useast.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000283568;r=20:29297095-29297246;t=ENST00000637374)

No reads found aligning there (three total across all samples)

```bash
for f in results/alignments/Homo_sapiens.GRCh38.dna.primary_assembly/sorted/*.bam; do echo $f; samtools view $f 20:29,297,095-29,297,246; done
```
