---
title: "Adamson smallRNA - pegRNA"
output:
  html_notebook:
    toc: true
---

## Load libraries

This project uses [`renv`](https://rstudio.github.io/renv/articles/renv.html)
to keep track of installed packages. Install `renv` if not installed and load
dependencies with `renv::restore()`.

```r
install.packages("renv")
renv::restore()
```

```{r message=FALSE}
library(readr)
library(dplyr)
library(tidyr)
library(pheatmap)
library(tibble)
library(ggplot2)
library(stringr)
library(cowplot)
library(markdown)
library(RColorBrewer)
```


## Read data

1. Get list of samples

```{r}
idxstats_dir <- "results/samtools_idxstats/Homo_sapiens.GRCh38.dna.primary_assembly/"
samples <- str_remove(list.files(idxstats_dir), ".bam.idxstats")
samples
```

2. Read Samtools `idxstats` to get human coverage for normalization

Note, the values from `idxstats` include the total number of reads aligned are
not limited to uniquely aligned reads.

```{r}
idxstats <- tibble(sample = samples, grch38_mapped_reads = NaN)
for (sample in samples) {

  # Read `idxstats` for normalization
  stats <- read_tsv(
    file.path(idxstats_dir, sprintf("%s.bam.idxstats", sample)),
    col_names = c(
      "sequence_name", "sequence_length",
      "mapped_reads", "unmapped_reads"
    ),
    col_types = "ciii"
  )
  grch38_mapped_reads <- stats %>%
    filter(!sequence_name %in% c("*")) %>%
    select(mapped_reads) %>%
    sum()
  idxstats <- rows_update(idxstats,
    tibble(
      sample = sample,
      grch38_mapped_reads = grch38_mapped_reads
    ),
    by = "sample"
  )
}
idxstats
```

3. Read `bedpe` files to get exogenous rna coverage of paired reads

```{r}
bedpe_data <- tibble()
for (sample in samples) {
  data <-
    readr::read_tsv(sprintf(
      "results/alignments/exogenous_rna/bedpe/%s.bedpe", sample
    ),
    col_names = c(
      "chrom1", "chrom1Start", "chrom1End",
      "chrom2", "chrom2Start", "chrom2End",
      "name", "score", "strand1", "strand2"
    ),
    col_types = "ciiciicicc"
    )
  bedpe_data <- tibble(rbind(
    bedpe_data,
    cbind(
      filter(idxstats, sample == !!sample),
      data
    )
  ))
}
bedpe_data
```

## Exogenous RNA annotation data

The exogenous RNA references have the following structure

**PJY103**
| start | end | description |
| --- | --- | --- |
| 1 | 330 | plasmid-1 |
| 331 | 351 | protospacer |
| 352 | 427 | gRNA scaffold |
| 428 | 453 | extension |
| 454 | 460 | terminator |
| 461 | 2306 | plasmid-2 |

Setup annotation for exogenous RNA regions

```{r}
annotation_pjy103 <- tibble(chrom2End = 1:2306, label = "plasmid")
annotation_pjy103[331:351, ]$label <- "protospacer"
annotation_pjy103[352:427, ]$label <- "gRNA scaffold"
annotation_pjy103[428:453, ]$label <- "extension"
annotation_pjy103[454:460, ]$label <- "terminator"
annotation_pjy103 <- annotation_pjy103 %>%
  mutate(region = paste0("PJY103_", chrom2End)) %>%
  select(-chrom2End) %>%
  column_to_rownames("region")
```

**PJY300**
| start | end | description |
| --- | --- | --- |
| 1 | 330 | plasmid-1 |
| 331 | 351 | protospacer |
| 352 | 427 | gRNA scaffold |
| 428 | 453 | extension |
| 454 | 490 | riboswitch |
| 491 | 497 | terminator |
| 498 | 2343 | plasmid-2 |

```{r}

annotation_pjy300 <- tibble(chrom2End = 1:2306, label = "plasmid")
annotation_pjy300[331:351, ]$label <- "protospacer"
annotation_pjy300[352:427, ]$label <- "gRNA scaffold"
annotation_pjy300[428:453, ]$label <- "extension"
annotation_pjy300[454:490, ]$label <- "riboswitch"
annotation_pjy300[491:497, ]$label <- "terminator"
annotation_pjy300 <- annotation_pjy300 %>%
  mutate(region = paste0("PJY300_", chrom2End)) %>%
  select(-chrom2End) %>%
  column_to_rownames("region")
```

## Summarize "cpmplete" fragments

"Complete" fragments are defined as fragments that start at or near the
protospacer. The following rules are used to generate counts for where the
sequenced fragment ends:

* Count only fragments that start within 2 bp of the (e)pegRNA
* Sum for each "end" position
* Normalize using reads mapped to human genome (grch38)

```{r}
summarized_data <- bedpe_data %>%
  filter(chrom1Start >= 329 & chrom1Start <= 333) %>%
  count(sample, grch38_mapped_reads, chrom1, chrom2End,
    name = "count"
  ) %>%
  mutate(normalizedCount = (count * 10^6) / grch38_mapped_reads) %>%
  # Fill in missing values between pegRNA start and end with 0
  complete(
    nesting(sample, grch38_mapped_reads, chrom1),
    chrom2End = seq(329, 501),
    fill = list(count = 0, normalizedCount = 0)
  )
summarized_data
```

Pivot data for correlation analysis

```{r}
corr_data <- summarized_data %>%
  select(-grch38_mapped_reads, -count) %>%
  pivot_wider(names_from = sample, values_from = normalizedCount) %>%
  arrange(chrom1, chrom2End)
corr_data
```

Save summarized data tables to text files

```{r}
write_tsv(summarized_data, file = "exogenous_targets_paired_read_summary.tsv")
write_tsv(corr_data, file = "exogenous_targets_normalized_counts.tsv")
```


## Transform data

Split data by pegRNA species and log transform

```{r}
heatmap_data_pjy103 <- corr_data %>%
  select("chrom1" | "chrom2End" | contains("PJY103")) %>%
  na.omit() %>%
  unite(region, c("chrom1", "chrom2End")) %>%
  mutate_if(is.numeric, log1p) %>%
  column_to_rownames("region")
heatmap_data_pjy103
```

```{r}
heatmap_data_pjy300 <- corr_data %>%
  select("chrom1" | "chrom2End" | contains("PJY300")) %>%
  na.omit() %>%
  unite(region, c("chrom1", "chrom2End")) %>%
  mutate_if(is.numeric, log1p) %>%
  column_to_rownames("region")
heatmap_data_pjy300
```

## Replicate correlation

Read in the sample metadata

```{r fig.height=16, fig.width=8}
samples <- read_tsv("config/samples.tsv")
units <- read_tsv("config/units.tsv")

annotation <- samples %>%
  left_join(units) %>%
  unite(sample, c(sample_name, unit_name)) %>%
  select(sample, cell_line, exogenous_rna, day) %>%
  column_to_rownames("sample") %>%
  mutate(across(c(cell_line, exogenous_rna, day), factor))

colors <- colorRampPalette(brewer.pal(9, "Blues"))(255)

newCols <- colorRampPalette(grDevices::rainbow(length(unique(annotation_pjy300$label))))
colors_pjy300 <- newCols(length(unique(annotation_pjy300$label)))
names(colors_pjy300) <- unique(annotation_pjy300$label)
colors_pjy300 <- list(label = colors_pjy300)
colors_pjy103 <- list(label = mycolors$label[names(mycolors$label) != "riboswitch"])

```





Heatmaps based on fragment lengths for each pegRNA species

```{r fig.height=12, fig.width=12}
heatmap_pjy103 <- pheatmap(heatmap_data_pjy103,
  cluster_rows = FALSE,
  show_rownames = FALSE,
  cluster_cols = TRUE,
  annotation_col = annotation %>% filter(exogenous_rna == "PJY103") %>% select(-exogenous_rna),
  annotation_row = annotation_pjy103,
  annotation_colors = colors_pjy103,
  main = "PJY103 Fragment Lengths",
  color = colors,
  silent = TRUE
)

heatmap_pjy300 <- pheatmap(heatmap_data_pjy300,
  cluster_rows = FALSE,
  show_rownames = FALSE,
  cluster_cols = TRUE,
  annotation_col = annotation %>% filter(exogenous_rna == "PJY300") %>% select(-exogenous_rna),
  annotation_row = annotation_pjy300,
  annotation_colors = colors_pjy300,
  main = "PJY300 Fragment Lengths",
  color = colors,
  silent = TRUE
)

cowplot::plot_grid(heatmap_pjy103$gtable, heatmap_pjy300$gtable, ncol = 2)
```


```{r fig.height=10, fig.width=12}
sample_dists <- dist(t(heatmap_data))
sample_dist_matrix <- as.matrix(sample_dists)
pheatmap(sample_dist_matrix,
  clustering_distance_rows = sample_dists,
  clustering_distance_cols = sample_dists,
  col = colors,
  annotation = annotation
)
```


---

## Exogenous species comparison

Heatmap functions
```{r}
blue_colors <- colorRampPalette(brewer.pal(9, "Blues"))(200)
diverging_colors <- colorRampPalette(brewer.pal(9, "RdBu"))(200)

sgrna_heatmap <- function(data, color = blue_colors, ...) {
  hmap <- pheatmap(
    data,
    cluster_rows = FALSE,
    color = color,
    fontsize_row = 6,
    ...
  )
  return(hmap)
}

filter_and_fill_data <- function(data, chrom, length) {
  filtered_data <- data %>%
    filter(chrom1 == chrom) %>%
    complete(
      nesting(chrom1, chrom1Start),
      chrom2End = seq(0, length),
      fill = list(count = 0, normalizedCount = 0)
    ) %>%
    unite(seqid, chrom1, chrom1Start, chrom2End) %>%
    column_to_rownames("seqid")
  return(filtered_data)
}


# Collapse Replicates
collapse_data <- function(data) {
  collapsed_data <- data %>%
    mutate(PJY142_PJY209 = rowMeans(
      select(., starts_with("PJY142_PJY209"))
    )) %>%
    mutate(PJY142_PJY99 = rowMeans(
      select(., starts_with("PJY142_PJY99"))
    )) %>%
    mutate(PJY151_PJY209 = rowMeans(
      select(., starts_with("PJY151_PJY209"))
    )) %>%
    mutate(PJY151_PJY99 = rowMeans(
      select(., starts_with("PJY151_PJY99"))
    )) %>%
    select(-contains("_rep"))
  return(collapsed_data)
}

# Scale
scale_data <- function(data, center = TRUE, scale = TRUE) {
  scaled_data <- data %>%
    mutate_at(
      vars(starts_with("PJY")),
      ~ (. / sum(., na.rm = TRUE))
      # ~ scale(., center = center, scale = scale)
    )
  return(scaled_data)
}
```


### PJY142

Histogram of fragments starting at position zero (0)
```{r, fig.width=12}
pjy142_length <- 113
pjy142_hist <-
  ggplot(
    summarized_data %>%
      filter(chrom1 == "PJY142") %>%
      complete(
        nesting(sample, grch38_mapped_reads, chrom1, chrom1Start),
        chrom2End = seq(0, pjy142_length),
        fill = list(count = 0, normalizedCount = 0)
      ),
    aes(x = chrom2End, y = normalizedCount, fill = sample)
  ) +
  geom_bar(stat = "identity") +
  theme(legend.position = "none") +
  facet_grid(vars(sample))
plot_grid(pjy142_hist, pjy142_hist + scale_y_continuous(trans = "log1p"))
```

#### Heatmaps PJY142

```{r, fig.width=12, fig.height=14}
pjy142_hmap <- sgrna_heatmap(
  corr_data %>%
    filter_and_fill_data("PJY142", pjy142_length) %>%
    log1p(),
  silent = TRUE
)
pjy142_hmap_collapsed <- sgrna_heatmap(
  corr_data %>%
    filter_and_fill_data("PJY142", pjy142_length) %>%
    collapse_data() %>%
    log1p(),
  silent = TRUE
)
plot_grid(pjy142_hmap$gtable, pjy142_hmap_collapsed$gtable)
```

Heatmaps with columns scaled by number of reads mapped to the region plotted
```{r, fig.width=12, fig.height=14}
pjy142_hmap_scaled <- sgrna_heatmap(
  corr_data %>%
    filter_and_fill_data("PJY142", pjy142_length) %>%
    scale_data(),
  silent = TRUE
)
pjy142_hmap_scaled_collapsed <- sgrna_heatmap(
  corr_data %>%
    filter_and_fill_data("PJY142", pjy142_length) %>%
    scale_data() %>%
    collapse_data(),
  silent = TRUE
)
plot_grid(pjy142_hmap_scaled$gtable, pjy142_hmap_scaled_collapsed$gtable)
```




### PJY151

Histogram of fragments starting at position zero (0)
```{r, fig.width=12}
pjy151_length <- 113
pjy151_hist <-
  ggplot(
    summarized_data %>%
      filter(chrom1 == "PJY151") %>%
      complete(
        nesting(sample, grch38_mapped_reads, chrom1, chrom1Start),
        chrom2End = seq(0, pjy151_length),
        fill = list(count = 0, normalizedCount = 0)
      ),
    aes(x = chrom2End, y = normalizedCount, fill = sample)
  ) +
  geom_bar(stat = "identity") +
  theme(legend.position = "none") +
  facet_grid(vars(sample))
plot_grid(pjy151_hist, pjy151_hist + scale_y_continuous(trans = "log1p"))
```

Heatmap PJY151
```{r, fig.width=12, fig.height=14}
pjy151_hmap <- sgrna_heatmap(
  corr_data %>%
    filter_and_fill_data("PJY151", pjy151_length) %>%
    log1p(),
  silent = TRUE
)
pjy151_hmap_collapsed <- sgrna_heatmap(
  corr_data %>%
    filter_and_fill_data("PJY151", pjy151_length) %>%
    collapse_data() %>%
    log1p(),
  silent = TRUE
)
plot_grid(pjy151_hmap$gtable, pjy151_hmap_collapsed$gtable)
```




```{r, fig.width=12, fig.height=14}
pjy151_hmap_scaled <- sgrna_heatmap(
  corr_data %>%
    filter_and_fill_data("PJY151", pjy151_length) %>%
    scale_data(),
  silent = TRUE
)
pjy151_hmap_scaled_collapsed <- sgrna_heatmap(
  corr_data %>%
    filter_and_fill_data("PJY151", pjy151_length) %>%
    scale_data() %>%
    collapse_data(),
  silent = TRUE
)
plot_grid(pjy151_hmap_scaled$gtable, pjy151_hmap_scaled_collapsed$gtable)
```


### PJY99

Histogram of fragments starting at position zero (0)
```{r, fig.width=12}
pjy99_length <- 138
pjy99_hist <-
  ggplot(
    summarized_data %>%
      filter(chrom1 == "PJY209", chrom1Start == 0) %>%
      filter(grepl("PJY99", sample)) %>%
      filter(chrom2End <= pjy99_length) %>%
      complete(
        nesting(sample, grch38_mapped_reads, chrom1, chrom1Start),
        chrom2End = seq(0, pjy99_length),
        fill = list(count = 0, normalizedCount = 0)
      ),
    aes(x = chrom2End, y = normalizedCount, fill = sample)
  ) +
  geom_bar(stat = "identity") +
  theme(legend.position = "none") +
  facet_grid(vars(sample))
plot_grid(pjy99_hist, pjy99_hist + scale_y_continuous(trans = "log1p"))
```


Heatmap PJY99
```{r, fig.width=12, fig.height=14}
pjy99_hmap <- sgrna_heatmap(
  corr_data %>%
    filter_and_fill_data("PJY209", pjy99_length) %>%
    select(!contains("PJY209")) %>%
    log1p(),
  silent = TRUE
)
pjy99_hmap_collapsed <- sgrna_heatmap(
  corr_data %>%
    filter_and_fill_data("PJY209", pjy99_length) %>%
    collapse_data() %>%
    select(!contains("PJY209")) %>%
    log1p(),
  silent = TRUE
)
plot_grid(pjy99_hmap$gtable, pjy99_hmap_collapsed$gtable)
```

```{r, fig.width=12, fig.height=16}
pjy99_hmap_scaled <- sgrna_heatmap(
  corr_data %>%
    filter_and_fill_data("PJY209", pjy99_length) %>%
    select(!contains("PJY209")) %>%
    scale_data(),
  silent = TRUE
)
pjy99_hmap_scaled_collapsed <- sgrna_heatmap(
  corr_data %>%
    filter_and_fill_data("PJY209", pjy99_length) %>%
    scale_data() %>%
    collapse_data() %>%
    select(!contains("PJY209")),
  silent = TRUE
)
plot_grid(pjy99_hmap_scaled$gtable, pjy99_hmap_scaled_collapsed$gtable)
```


### PJY209

Histogram of fragments starting at position zero (0)
```{r, fig.width=12}
pjy209_length <- 174
pjy209_hist <-
  ggplot(
    summarized_data %>%
      filter(chrom1 == "PJY209", chrom1Start == 0) %>%
      filter(grepl("PJY209", sample)) %>%
      complete(
        nesting(sample, grch38_mapped_reads, chrom1, chrom1Start),
        chrom2End = seq(0, pjy209_length),
        fill = list(count = 0, normalizedCount = 0)
      ),
    aes(x = chrom2End, y = normalizedCount, fill = sample)
  ) +
  geom_bar(stat = "identity") +
  theme(legend.position = "none") +
  facet_grid(vars(sample))
plot_grid(pjy209_hist, pjy209_hist + scale_y_continuous(trans = "log1p"))
```

Heatmap PJY209
```{r, fig.width=12, fig.height=16}
pjy209_hmap <- sgrna_heatmap(
  corr_data %>%
    filter_and_fill_data("PJY209", pjy209_length) %>%
    select(contains("PJY209")) %>%
    log1p(),
  silent = TRUE
)
pjy209_hmap_collapsed <- sgrna_heatmap(
  corr_data %>%
    filter_and_fill_data("PJY209", pjy209_length) %>%
    collapse_data() %>%
    select(contains("PJY209")) %>%
    log1p(),
  silent = TRUE
)
plot_grid(pjy209_hmap$gtable, pjy209_hmap_collapsed$gtable)
```

```{r, fig.width=12, fig.height=16}
pjy209_hmap_scaled <- sgrna_heatmap(
  corr_data %>%
    filter_and_fill_data("PJY209", pjy209_length) %>%
    select(contains("PJY209")) %>%
    scale_data(),
  silent = TRUE,
)
pjy209_hmap_scaled_collapsed <- sgrna_heatmap(
  corr_data %>%
    filter_and_fill_data("PJY209", pjy209_length) %>%
    scale_data() %>%
    collapse_data() %>%
    select(contains("PJY209")),
  silent = TRUE,
)
plot_grid(pjy209_hmap_scaled$gtable, pjy209_hmap_scaled_collapsed$gtable)
```

```{r}
pjy209_scaled_diff <-
  corr_data %>%
  filter_and_fill_data("PJY209", pjy209_length) %>%
  scale_data() %>%
  collapse_data() %>%
  select(contains("PJY209")) %>%
  mutate(PJY_DIFF = PJY151_PJY209 - PJY142_PJY209)
pjy209_scaled_diff
```

#### Heatmap of Difference

Plot the difference between the scaled reads for PJY151 vs PJY142

```{r, fig.width=12, fig.height=16}
pjy209_hmap_comp <- sgrna_heatmap(
  corr_data %>%
    filter_and_fill_data("PJY209", pjy209_length) %>%
    scale_data() %>%
    collapse_data() %>%
    select(contains("PJY209")),
  silent = TRUE,
  treeheight_col = 0,
  show_colnames = FALSE,
  main = "Scaled PJY151 and PJY142"
)

pjy209_hmap_diff <- sgrna_heatmap(
  pjy209_scaled_diff %>% select("PJY_DIFF"),
  cluster_cols = FALSE,
  silent = TRUE,
  color = diverging_colors,
  scale = "column",
  show_colnames = FALSE,
  main = "PJY151 vs PJY142"
)

plot_grid(pjy209_hmap_comp$gtable,
  pjy209_hmap_diff$gtable,
  rel_widths = c(2, 1)
)
```
